crates/runc-shim/src/cgroup_memory.rs

```diff
@@ -0,0 +1,126 @@
+use containerd_shim::{
+    error::{Error, Result},
+    io_error, other_error,
+};
+use std::path::Path;
+use tokio::{
+    fs::{self, File},
+    io::{AsyncBufReadExt, AsyncReadExt, BufReader},
+    sync::mpsc::{self, Receiver},
+};
+
+use nix::sys::eventfd;
+use std::os::unix::io::{AsRawFd, FromRawFd};
+
+pub async fn get_path_from_cgorup(pid: u32) -> Result<String> {
+    let proc_path = format!("/proc/{}/cgroup", pid);
+    let file = File::open(proc_path.clone())
+        .await
+        .map_err(io_error!(e, "open {}.", proc_path))?;
+    let mut lines = BufReader::new(file).lines();
+    while let Some(line) =
+        lines
+            .next_line()
+            .await
+            .map_err(io_error!(e, "read file {} failed.", proc_path))?
+    {
+        if line.contains("memory") {
+            if let Some(path) = line.split(':').last() {
+                return Ok(path.to_string());
+            }
+            break;
+        }
+    }
+    Err("Memory cgroup path not found").map_err(other_error!(e, "Error get path:"))?
+}
+
+pub async fn get_existing_cgroup_mem_path(pid_path: String) -> Result<(String, String)> {
+    let mut mountinfo_path = get_path_from_mountinfo().await?;
+    if mountinfo_path.0 == "/" {
+        mountinfo_path.0 = String::from("");
+    }
+    let res = pid_path.trim_start_matches(&mountinfo_path.0).to_string();
+    Ok((res, mountinfo_path.1))
+}
+
+async fn get_path_from_mountinfo() -> Result<(String, String)> {
+    let mountinfo_file = File::open("/proc/self/mountinfo".clone())
+        .await
+        .map_err(io_error!(e, "open {}", "/proc/self/mountinfo"))?;
+
+    // for line in BufReader::new(mountinfo_file).lines() {
+    let mut lines = BufReader::new(mountinfo_file).lines();
+    while let Some(line) = lines
+        .next_line()
+        .await
+        .map_err(io_error!(e, "read file /proc/self/mountinfo failed."))?
+    {
+        if line.contains("cgroup") && line.contains("memory") {
+            return parse_memory_mountroot(&line);
+        }
+    }
+
+    Err("No memory found in mountinfo file.")
+        .map_err(other_error!(e, "Error get path from montinfo:"))?
+}
+
+fn parse_memory_mountroot(line: &str) -> Result<(String, String)> {
+    let s_values = line
+        .split_once(" - ")
```

Comment:
I don't think we need to split the line by `-`, and then by whitespace.
We're interested in the first block anyway, so something like this should do?


```rust
    let mut lines = line.split_whitespace();

    let root = lines.nth(3).ok_or(Error::Other("...".into()))?;
    let mount = lines.next(). ok_or(Error::Other("...".into()))?;

    return (root.to_string(), mount.to_string())
```

Reply:
The reson i used `-` to split the sting is because i referred to some libs for parsing the mountinfo file, using it to verify the validation of information about the mountinfo file. But i found it unnecessary using `-` after checking different platforms and compared with go shim, so here i will make some modifications here.